File: code/PubSub/src/main/scala/SubMW.scala
import com.rabbitmq.client.{AMQP, Channel, Connection, ConnectionFactory, DefaultConsumer, Envelope, MessageProperties}

object SubMW {
  private var connection: Connection = _
  private var channel: Channel = _
  private var queueName: String = _

  def connect_to_server(host: String, port: Int): Unit = {
    val factory = new ConnectionFactory()
    factory.setHost(host)
    factory.setPort(port)
    connection = factory.newConnection() // Corrected method call
    channel = connection.createChannel()
  }

  def setup_subscriber_queue(exchangeName: String, callback: String => Unit): Unit = {
    queueName = channel.queueDeclare().getQueue
    channel.queueBind(queueName, exchangeName, "")

    val consumer = new DefaultConsumer(channel) {
      override def handleDelivery(consumerTag: String, envelope: Envelope, properties: AMQP.BasicProperties, body: Array[Byte]): Unit = {
        try {
          val message = new String(body, "UTF-8")
          callback(message)
        } catch {
          case e: Exception =>
            println(s"Error while processing message: ${e.getMessage}")
        }
      }
    }

    channel.basicConsume(queueName, true, consumer)
  }
}

File: PubSubRaft/src/main/kotlin/SubMW.kt
package org.example

import com.rabbitmq.client.*
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.fasterxml.jackson.module.kotlin.readValue

interface SubUpcallHandler {
    fun onMessageReceived(message: String)
    fun onErrorOccurred(error: Throwable)
}

object SubMW {
    private var connection: Connection? = null
    private var channel: Channel? = null
    private var queueName: String? = null
    private var upcallHandler: SubUpcallHandler? = null

    private val objectMapper: ObjectMapper = jacksonObjectMapper()

    fun connectToServer(host: String, port: Int) {
        val factory = ConnectionFactory()
        factory.host = host
        factory.port = port
        connection = factory.newConnection()
        channel = connection?.createChannel()
    }

    fun setupSubscriberQueue(exchangeName: String) {
        queueName = channel?.queueDeclare()?.queue
        channel?.queueBind(queueName, exchangeName, "")

        val consumer = object : DefaultConsumer(channel) {
            override fun handleDelivery(
                consumerTag: String,
                envelope: Envelope,
                properties: AMQP.BasicProperties,
                body: ByteArray
            ) {
                try {
                    val message = String(body, Charsets.UTF_8)
                    val deserializedMessage: String = objectMapper.readValue(message)
                    upcallHandler?.onMessageReceived(deserializedMessage)
                } catch (e: Exception) {
                    upcallHandler?.onErrorOccurred(e)
                }
            }
        }

        channel?.basicConsume(queueName, true, consumer)
    }

    fun setUpcallHandler(handler: SubUpcallHandler) {
        upcallHandler = handler
    }

    fun unsubscribe(exchangeName: String) {
        if (channel != null && queueName != null) {
            channel?.queueUnbind(queueName, exchangeName, "")
        }
    }

    fun changeSubscription(newExchangeName: String) {
        unsubscribe("")
        setupSubscriberQueue(newExchangeName)
    }

    fun disconnect() {
        if (channel != null && queueName != null) {
            channel?.queueDelete(queueName)
            channel?.close()
        }
        connection?.close()
    }
}
